### PR TITLE
Using Philips B0 fieldmaps for distortion correction of structural and functional images

### PR DESCRIPTION
**Pull request by sivcek**
Hi,

This is a pull request that adds the ability to use Philips `B0-map` field map image generated by Philips scanners for distortion correction of both structural as well as functional images. The code has been tested successfully on both structural and functional images from a Philips scanner. It does not change any other parts of the processing pipelines.

With kind regards,

Grega & Aleksij

### CODE REVIEW

File: PreFreeSurfer/PreFreeSurferPipeline.sh [hunk 1]
```diff
@@ -189,8 +190,8 @@ Usage: ${script_name} [options]
   --template2mmmask=<file path>       Brain mask MNI 2mm Template
   --brainsize=<size value>            Brain size estimate in mm, 150 for humans
   --fnirtconfig=<file path>           FNIRT 2mm T1w Configuration file
-  --fmapmag=<file path>               Siemens Gradient Echo Fieldmap magnitude file
-  --fmapphase=<file path>             Siemens Gradient Echo Fieldmap phase file
+  --fmapmag=<file path>               Siemens/Philips Gradient Echo Fieldmap magnitude file
```
Comment (glasserm) [inline PreFreeSurfer/PreFreeSurferPipeline.sh:193]: Why not new arguments?  Is a Philips fieldmap formatted the same as a Siemens fieldmap?

**Comment by glasserm** [comment]
This code looks like it borrowed very heavily from the Siemens field map code.  What are the differences between a Siemens and Philips fieldmap?

File: PreFreeSurfer/PreFreeSurferPipeline.sh (hunk 1 above)
Comment (axkralj990) [reply PreFreeSurfer/PreFreeSurferPipeline.sh:193]: Yes, both Philips and Siemens field map scans consist of two separate MRI images, one for the magnitude data and the other for phase data. In contrast, General Electric field map scans have magnitude and phase data stored in two volumes of a single image.

**Comment by axkralj990** [comment]
Here are some of the notes I wrote while working on the Philips gradient-echo (GRE) fieldmap distortion correction (DC). The differences between Philips and Siemens fieldmap DC are described under point 1.

1. We have written a shell script that performs field map DC on structural and functional scans - `${HCPPIPEDIR}/global/scripts/PhilipsFieldMapPreprocessingAll.sh`
   The script was largely based on the existing `SiemensFieldMapPreprocessingAll.sh` and `GeneralElectricFieldMapPreprocessingAll.sh` scripts. Specific information on performing DC using Philips B0 (GRE field map) scans was obtained from the document by *Anna van 't Veer, Anne Hafkemeijer, Henk van Steenbergen and Janna Marie Bas-Hoogendam (Leiden University / Leiden Institute for Brain and Cognition)*: https://osf.io/hks7x/ (recommended by FSL's fugue documentation for non-Siemens data (https://fsl.fmrib.ox.ac.uk/fsl/fslwiki/FUGUE/Guide)). This procedure does not use the echo difference, as opposed to Siemens DC procedure implemented in FSL's function `fsl_prepare_fieldmap`, which currently only supports Siemens data. The echo difference is used as `asym` parameter for FSL's `fugue` function, but was excluded from the `fugue` call in `PhilipsFieldMapPreprocessingAll.sh`, because we strictly followed the recommended steps for DC on Philips data, where they did not include `asym` when calling `fugue`.

   Philips and Siemens field map scans consist of two separate MRI images, one for the magnitude data and the other for phase data. In contrast, General Electric field map scans have magnitude and phase data stored in two volumes of a single image.

   Shown below are the code sections from the `global/scripts/*FieldMapPreprocessingAll.sh` scripts, which show the main differences between the data from different scanners.

   **PhilipsFieldMapPreprocessingAll.sh - lines 108-118:**

   ```bash
   108 ${FSLDIR}/bin/fslmaths ${MagnitudeInputName} -Tmean ${WD}/Magnitude
   109 ${FSLDIR}/bin/bet ${WD}/Magnitude ${WD}/Magnitude_brain -f 0.35 -m #Brain extract the magnitude image
   110 ${FSLDIR}/bin/fslmaths ${WD}/Magnitude_brain -ero ${WD}/Magnitude_brain_ero
   111 rm ${WD}/Magnitude_brain.nii.gz
   112 mv ${WD}/Magnitude_brain_ero.nii.gz ${WD}/Magnitude_brain.nii.gz
   113 
   114 ${FSLDIR}/bin/imcp ${PhaseInputName} ${WD}/FieldMap_deg
   115 ${FSLDIR}/bin/fslmaths ${WD}/FieldMap_deg -mul 6.28 ${WD}/FieldMap_rad
   116 ${FSLDIR}/bin/fugue --loadfmap=${WD}/FieldMap_rad.nii.gz -m --savefmap=${WD}/FieldMap.nii.gz
   117 
   118 log_Msg "DONE: fsl_prepare_fieldmap.sh"
   ```

   **SiemensFieldMapPreprocessingAll.sh - lines 110-115:**

   ```bash
   110 ${FSLDIR}/bin/fslmaths ${MagnitudeInputName} -Tmean ${WD}/Magnitude
   111 ${FSLDIR}/bin/bet ${WD}/Magnitude ${WD}/Magnitude_brain -f 0.35 -m #Brain extract the magnitude image
   112 ${FSLDIR}/bin/imcp ${PhaseInputName} ${WD}/Phase
   113 ${FSLDIR}/bin/fsl_prepare_fieldmap SIEMENS ${WD}/Phase ${WD}/Magnitude_brain ${WD}/FieldMap ${DeltaTE}
   114 
   115 log_Msg "DONE: fsl_prepare_fieldmap.sh"
   ```

   **GeneralElectricPreprocessingAll.sh - line 106-112:**

   ```bash
   106 ${FSLDIR}/bin/fslsplit ${GEB0InputName}     # split image into vol0000=fieldmap and vol0001=magnitude
   107 mv vol0000.nii.gz ${WD}/FieldMap_deg.nii.gz
   108 mv vol0001.nii.gz ${WD}/Magnitude.nii.gz
   109 ${FSLDIR}/bin/bet ${WD}/Magnitude ${WD}/Magnitude_brain -f 0.35 -m #Brain extract the magnitude image
   110 ${FSLDIR}/bin/fslmaths ${WD}/FieldMap_deg.nii.gz -mul 6.28 ${WD}/FieldMap.nii.gz
   111 
   112 log_Msg "DONE: preparing General Electric fieldmap"
   ```

2. Philips GRE field map DC was then implemented for uwarping T1w and T2w scans during the PreFreeSurfer hcp preprocessing step by modifying the following 2 functions:

   * `${HCPPIPEDIR}/PreFreeSurfer/scripts/T2wToT1wDistortionCorrectAndReg.sh`
   * `${HCPPIPEDIR}/PreFreeSurfer/PreFreeSurferPipeline.sh`

3. Finally, we included Philips GRE field map DC for unwarping functional scans during the fMRIVolume hcp preprocessing step, by modifying the following two functions:

   * `${HCPPIPEDIR}/fMRIVolume/GenericfMRIVolumeProcessingPipeline.sh`
   * `${HCPPIPEDIR}/fMRIVolume/scripts/DistortionCorrectionAndEPIToT1wReg_FLIRTBBRAndFreeSurferBBRbased.sh`

**Comment by glasserm** [comment]
Given these relatively small differences between fieldmaps from different vendors, I wonder if it would be better to have a single fieldmap processing script with cases for the differences?

**Comment by sivcek** [comment]
@glasserm This would require a bit more thorough rewrite and testing. The code as is already works, and it does support more flexible handling of different fieldmaps.

**Comment by glasserm** [comment]
I'm not insisting on it, but maintenance would be greatly improved if there were a single script.  In any case, I am happy to review some results.

**Comment by sivcek** [comment]
@glasserm, @axkralj990 is finalizing results to review and will let you know once prepared. In the meantime, we can investigate how complicated it would be to unify and test the joined scripts. They would still need to be separate from the SE based script. Do you have a suggestion for a common name? E.g. `EPIFieldMapPreprocessingAll.sh`?

**Comment by glasserm** [comment]
I would get rid of "EPI" since none of these fieldmaps are EPI-based and they are used for correcting more than EPI data.

**Comment by glasserm** [comment]
Out of date

**Pull request closed without merging**